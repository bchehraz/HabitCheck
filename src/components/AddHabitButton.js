import React, { useState } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { FaPlus } from 'react-icons/fa';

const ButtonContainer = styled.div`
  margin-bottom: 14px;
  padding: 10px;
  width: 327px;
  height: ${({ formOpen }) => (formOpen) ? "180px" : "53px"};

  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: flex-start;

  border: 2px solid #2D3142;
  border-radius: 17px;
  background-color: #2D3142;
  color: white;

  -webkit-transition: all 300ms ease-out;
  -moz-transition: all 300ms ease-out;
  -ms-transition: all 300ms ease-out;
  -o-transition: all 300ms ease-out;
  transition: all 300ms ease-out;

  .iconContainerClose {
    -moz-transform: translateX(120px);
    -webkit-transform: translateX(120px);
    -o-transform: translateX(120px);
    -ms-transform: translateX(120px);
    transform: translateX(120px);
  }
`;

const IconContainer = styled.div`
  padding: 10px;
  padding: ${({ formOpen }) => formOpen && '5px 10px' || 0};
  position: absolute;;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;

  -webkit-transition: all 300ms ease-out;
  -moz-transition: all 300ms ease-out;
  -ms-transition: all 300ms ease-out;
  -o-transition: all 300ms ease-out;
  transition: all 300ms ease-out;

  .icon {
    -webkit-transition: all 300ms ease-out;
    -moz-transition: all 300ms ease-out;
    -ms-transition: all 300ms ease-out;
    -o-transition: all 300ms ease-out;
    transition: all 300ms ease-out;
  }

  .iconClose {
    -moz-transform: rotate(45deg);
    -webkit-transform: rotate(45deg);
    -o-transform: rotate(45deg);
    -ms-transform: rotate(45deg);
    transform: rotate(45deg);
  }
`;

const FormContainer = styled.form`
  margin-top: 2rem;
  padding: 10px 1rem;
  width: 100%;
  opacity: ${({ formOpen }) => formOpen ? 1 : 0};
  -webkit-transition: all ${({ formOpen }) => formOpen ? '600ms' : '300ms'} ease-out;
  -moz-transition: all ${({ formOpen }) => formOpen ? '600ms' : '300ms'} ease-out;
  -ms-transition: all ${({ formOpen }) => formOpen ? '600ms' : '300ms'} ease-out;
  -o-transition: all ${({ formOpen }) => formOpen ? '600ms' : '300ms'} ease-out;
  transition: all ${({ formOpen }) => formOpen ? '600ms' : '300ms'} ease-out;
`;

const TextInput = styled.input`
  padding: 10px 5px;
  width: 100%;
  margin: 0 auto;
  font-weight: 600;
  ${({ error }) => error && 'border: 1px solid red;'}
  ::placeholder {
    ${({ error }) => error && 'color: red;'}
  }
`;

const Button = styled.button`
  color: white;
  background-color: #8ACB88;
  font-weight: bold;
  width: 100%;
  padding: 10px;
  margin-top: 8px;
  border-radius: 6px;

  :active {
    outline: 1;
  }
`;

/*

  TODO: Figure out what is meant to be an onClick listener in this case.
  * For the sake of animating the ButtonContainer to an open state, the ButtonContainer must continue to be what is the container for the form. However, since the onClick container is the entire length/width of the form container once it's open, perhaps the onclick could just return false and only close if a new habit is added or if the X button is pushed

*/

  /*
    TODO: add a form which onClick will trigger a contextual function which will be passed down from parent component(s), either HabitList or Today...

    TODO: Figure out what functions need to be added into context and figure out if AuthContext should be changed out for UserContext which would include all user data in general.
    - functions could be;
      -
  */

const validForm = (formInput) => {
  const { title } = formInput;
  return title && title !== "";
}

const AddHabitButton = ({ handleAddHabit }) => {
  const initState = {
    title: "",
  }

  const [formOpen, setFormOpen] = useState(false);
  const [formInput, setFormInput] = useState(initState);
  const [error, setError] = useState(false);

  console.log(formInput.title);
  return (
    <ButtonContainer formOpen={formOpen} onClick={() => {
      if (!formOpen) {
        console.log("<Action> Add Habit Button was clicked!");
        setFormOpen(true);
      }
    }}>
      <FormContainer formOpen={formOpen}>
        <TextInput type="text"
          value={formInput.title}
          onChange={(e) => {
            setFormInput({ ...formInput, title: e.target.value });
          }}
          placeholder="Habit Title"
          error={error}
        />
        <Button
          onClick={(e) => {
            e.preventDefault();
            if (validForm(formInput)) {
              handleAddHabit(formInput.title);
              setError(false);
              setFormOpen(false);
              setFormInput(initState);
            } else {
              setError(true);
            }
          }}
        >
          {`Start New Habit`}
        </Button>
      </FormContainer>

      <IconContainer
        formOpen={formOpen}
        className={(formOpen) && "iconContainerClose"}
        onClick={() => {
          if (formOpen) {
            setFormOpen(false);
            setError(false);
            setFormInput(initState);
          }
        }}
      >
        <FaPlus
          size={25}
          className={(formOpen) && "icon iconClose" || "icon"}
        />
      </IconContainer>
    </ButtonContainer>
  );
}

AddHabitButton.propTypes = {
  handleAddHabit: PropTypes.func.isRequired,
}

export default AddHabitButton;
