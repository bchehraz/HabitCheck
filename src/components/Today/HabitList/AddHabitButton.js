import React, { useState, useRef } from "react"
import PropTypes from "prop-types"
import styled from "styled-components"
import { FaPlus } from "react-icons/fa"
import { BsCheck } from "react-icons/bs"

import Backdrop from "../../Backdrop"

const Container = styled.div`
  position: relative;
  margin: 0;
  padding: 0;
  width: 100%;
  height: ${({ formOpen }) => (formOpen ? "180px" : "53px")};

  display: flex;
  flex-flow: column nowrap;
  justify-content: center;
  align-items: center;

  border: 2px solid #2d3142;
  border-radius: 17px;
  background-color: #2d3142;
  color: white;

  transition: all 300ms ease-out;
  z-index: 10;

  @media only screen and (min-width: 450px) {
    height: ${({ formOpen }) => (formOpen ? "119px" : "53px")};
  }

  @media only screen and (min-width: 768px) {
    height: 119px;
    box-shadow: 3px 3px 7px rgba(0, 0, 0, 0.29);
    background: linear-gradient(97.52deg, #66c7f4 0%, #8acb88 100%);
    border: none;
  }

  @media only screen and (min-width: 1024px) {
    height: 135px;
  }

  @media only screen and (min-width: 1440px) {
    height: 150px;
  }
`

const IconContainer = styled.div`
  position: absolute;
  display: flex;
  flex-flow: row nowrap;
  justify-content: center;
  align-items: center;

  transition: all 300ms ease-out;
  opacity: ${props => (props.formOpen ? 0 : 1)};

  .icon {
    width: 25px;
    height: 25px;
  }

  @media only screen and (min-width: 768px) {
    .icon {
      width: 40px;
      height: 40px;
    }
  }
`

const FormContainer = styled.form`
  display: flex;
  flex-flow: column nowrap;
  z-index: 10;
  width: 100%;
  padding: 1rem;

  transition: all 150ms ease-out;
  margin: 0 auto;

  transform: translateY(75px);

  opacity: 0;
  visibility: hidden;

  ${props =>
    props.formOpen &&
    `
    visibility: visible;
    opacity: 1;
    transform: translateY(0);
    transform: translateX(0);
    transition: all 300ms ease-out;
  `}

  @media only screen and (min-width: 768px) {
    transform: translateY(0);
    ${props =>
      props.formOpen &&
      `
      visibility: visible;
      opacity: 1;
      transition: all 300ms ease-out;
    `}
  }
`

const FormSection = styled.div`
  display: flex;
  flex-flow: column nowrap;
  margin-top: 10px;

  @media only screen and (min-width: 450px) {
    flex-flow: row nowrap;
    justify-content: center;
    align-items: center;
  }
`

const TextInput = styled.input`
  padding: 10px 5px;
  width: 100%;
  background-color: #6f789d;
  color: white;
  border-bottom: 2px solid #8acb88;

  ${({ error }) =>
    error &&
    `
    border-bottom: 3px solid #AA4465;
  `}

  @media only screen and (min-width: 450px) {
    padding: 10px 10px;
    color: #2d3142;
    font-weight: 600;
  }

  @media only screen and (min-width: 768px) {
    font-size: 1.2em;
    background-color: white;
  }
`

const SubmitButton = styled.button`
  color: white;
  background-color: #8acb88;
  width: 100%;
  margin-top: 10px;
  padding: 5px;

  .buttonIcon {
    width: 30px;
    height: 30px;
  }

  @media only screen and (min-width: 450px) {
    width: auto;
    padding: 0;
    margin-top: 0;
    margin-left: 10px;
    background-color: rgba(0, 0, 0, 0);

    .buttonIcon {
      width: 40px;
      height: 100%;
      color: #8acb88;
    }
  }

  @media only screen and (min-width: 768px) {
    .buttonIcon {
      color: #fff;
    }
  }
`

const Header = styled.h4`
  color: white;
  margin: 0;
  text-align: left;
`

const validForm = formInput => {
  const { title } = formInput
  return title && title !== "" && title.length <= 64
}

const AddHabitButton = ({ handleAddHabit, onNewHabitAdded }) => {
  const initState = {
    title: "",
  }

  const [formOpen, setFormOpen] = useState(false)
  const [formInput, setFormInput] = useState(initState)
  const [error, setError] = useState(false)
  const textInput = useRef(null)

  const onFormOpen = () => {
    if (!formOpen) {
      setFormOpen(true)

      // Due to Form goes from hidden to visible,
      // input focus() won't work if the function is run immediately after
      // this is the solution for the time being
      window.setTimeout(() => {
        textInput.current.focus()
      }, 300)
    }
  }

  const onFormClose = () => {
    if (formOpen) {
      setFormOpen(!formOpen)
      setError(false)
      setFormInput(initState)
    }
  }

  const onTextChange = e => {
    const { value } = e.target
    setError(!(value.length <= 64))
    setFormInput({ ...formInput, title: e.target.value })
  }

  return (
    <div>
      <Container
        className="addHabitButton"
        formOpen={formOpen}
        onClick={onFormOpen}
      >
        <FormContainer formOpen={formOpen} className="form">
          <Header formOpen={formOpen}>{"New Habit"}</Header>

          <FormSection>
            <TextInput
              className="habitTitleInput"
              type="text"
              value={formInput.title}
              onChange={onTextChange}
              error={error}
              ref={textInput}
            />

            <SubmitButton
              onClick={e => {
                e.preventDefault()
                if (validForm(formInput)) {
                  handleAddHabit(formInput.title)
                  setError(false)
                  setFormOpen(false)
                  setFormInput(initState)
                  onNewHabitAdded()
                } else {
                  setError(true)
                }
              }}
            >
              <BsCheck className="buttonIcon" />
            </SubmitButton>
          </FormSection>
        </FormContainer>

        <IconContainer
          formOpen={formOpen}
          className={(formOpen && "iconContainerClose") || "iconContainerOpen"}
          data-testid="AddHabitButtonClose"
        >
          <FaPlus className="icon" />
        </IconContainer>
      </Container>
      <Backdrop onClick={onFormClose} enabled={formOpen} zIndex={9} />
    </div>
  )
}

AddHabitButton.propTypes = {
  handleAddHabit: PropTypes.func.isRequired,
  onNewHabitAdded: PropTypes.func.isRequired,
}

export default AddHabitButton
