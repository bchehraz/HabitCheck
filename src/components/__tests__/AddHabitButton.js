import React from "react"
import { render, fireEvent } from "@testing-library/react"
// import renderer from "react-test-renderer"

import AddHabitButton from "../AddHabitButton"

test("temp", () => {
  expect(true).toBe(true)
})

test("AddHabitButton: handleAddHabit callback with input", () => {
  // pass in a function which increments variable i
  let i = 0
  const { getByText, debug, container, getByPlaceholderText } = render(
    <AddHabitButton
      handleAddHabit={() => {
        i = i + 1
      }}
    />
  )

  fireEvent.click(container)
  expect(getByText(/start new habit/i))

  fireEvent.change(getByPlaceholderText(/habit title/i), {
    target: { value: "test title" },
  })
  fireEvent.click(getByText(/start new habit/i))

  expect(i).toBe(1)
})

test("AddHabitButton: handleAddHabit callback without input", () => {
  let i = 0
  const { getByText, debug, container, getByPlaceholderText } = render(
    <AddHabitButton
      handleAddHabit={() => {
        i = i + 1
      }}
    />
  )

  fireEvent.click(container)
  expect(getByText(/start new habit/i))

  fireEvent.click(getByText(/start new habit/i))

  expect(i).toBe(0)
  //debug();
})

test("AddHabitButton: closing the window works", () => {
  let i = 0
  const { getByTestId, debug, getByText, container } = render(
    <AddHabitButton
      handleAddHabit={() => {
        i = i + 1
      }}
    />
  )

  fireEvent.click(container)
  expect(getByText(/start new habit/i))

  fireEvent.click(getByTestId(/AddHabitButtonClose/i))

  expect(i).toBe(0)
})
