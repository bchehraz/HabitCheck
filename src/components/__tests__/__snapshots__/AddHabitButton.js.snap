// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddHabitButton: Snapshot test 1`] = `
<div
  className="sc-bdVaJa drkGtX addHabitButton"
  onClick={[Function]}
>
  <form
    className="sc-htpNat bXqxRp"
  >
    <input
      className="sc-bxivhb gGYjOE habitTitleInput"
      onChange={[Function]}
      placeholder="Habit Title"
      type="text"
      value=""
    />
    <button
      className="sc-ifAKCX dsDEqc"
      onClick={[Function]}
    >
      Start New Habit
    </button>
  </form>
  <div
    className="sc-bwzfXH hofZqQ"
    data-testid="AddHabitButtonClose"
    onClick={[Function]}
  >
    <svg
      className="icon"
      fill="currentColor"
      height={25}
      size={25}
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 448 512"
      width={25}
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
      />
    </svg>
  </div>
</div>
`;
